reana_server/fetcher.py

```diff
@@ -0,0 +1,243 @@
+        :param spec_name: Filename of the workflow specification file to be fetched.
+        """
+        super().__init__(url, output_dir, spec_name)
+
+    def fetch(self) -> None:
+        """Fetch workflow specification from a given URL."""
+        workflow_spec_path = os.path.join(self._output_dir, self._spec)
+        urlretrieve(self._url, workflow_spec_path)
+
+
+class WorkflowFetcherZip(WorkflowFetcherBase):
+    """Fetch the specification of a workflow from a zip archive."""
+
+    def __init__(self, url: str, output_dir: str, archive_name: str):
+        """Initialize the workflow specification fetcher.
+
+        :param url: URL of the workflow specification to fetch.
+        :param output_dir: Directory where all the data will be saved to.
+        :param archive_name: Filename of the zip archive to be fetched.
+        """
+        super().__init__(url, output_dir)
+        self._archive_name = archive_name
+
+    def fetch(self) -> None:
+        """Fetch workflow specification from a zip archive."""
+        archive_path = os.path.join(self._output_dir, self._archive_name)
+        urlretrieve(self._url, archive_path)
+        with zipfile.ZipFile(archive_path, "r") as zip_file:
+            zip_file.extractall(path=self._output_dir)
+        os.remove(archive_path)
+
+        if not self._discover_workflow_specs():
+            top_level_entries = [
+                os.path.join(self._output_dir, entry)
+                for entry in os.listdir(self._output_dir)
+            ]
+            # Some zip archives contain a single directory with all the files.
+            if len(top_level_entries) == 1 and os.path.isdir(top_level_entries[0]):
+                top_level_dir = top_level_entries[0]
+                # Move all entries inside the top level directory
+                # to the output directory.
+                for entry in os.listdir(top_level_dir):
+                    shutil.move(os.path.join(top_level_dir, entry), self._output_dir)
+                os.rmdir(top_level_dir)
+
+
+def _get_github_fetcher(url: str, output_dir: str) -> WorkflowFetcherGit:
+    """Parse a GitHub URL and return the correct fetcher.
+
+    :param url: URL to a GitHub repository.
+    :param output_dir: Directory where all the data fetched will be saved.
+    :returns: Workflow fetcher.
+    """
+    # There are four different GitHub URLs:
+    # 1. URL to a repository: /<user>/<repo>
+    # 2. URL to a branch/commit: /<user>/<repo>/tree/<git_ref>
+    # 3. URL to a specific folder: /<user>/<repo>/tree/<git_ref>/<path_to_dir>
+    # 4. URL to a specific file: /<user>/<repo>/blob/<git_ref>/<path_to_file>
+
+    # We are interested in five components: username, repository, tree/blob, git_ref, path
+    NUM_COMPONENTS = 5
+    parsed_url = urlparse(url)
+    url_path_components = parsed_url.path.strip("/").split(
+        "/", maxsplit=NUM_COMPONENTS - 1
+    )
+    url_path_components = url_path_components + [None] * (
```

Comment:
OK, I understand now... IMHO I find it quite hacky and not very easy to read.. It might be worth trying to find another way of solving this or at least leave a comment.

Reply:
Not much better.. but I came up with this solution:

```python
def get_url_path_components():
  url_bits = parsed_url.path.strip("/").split("/", maxsplit=NUM_COMPONENTS - 1)
  for i in range(NUM_COMPONENTS):
      try:
          yield url_bits[i]
      except IndexError:
          yield None
```

Not very happy about it though, you can ignore if you don't like it.